docs/RFCS/replicate_queue_remove_replicas.md

```diff
@@ -0,0 +1,166 @@
+exotic failure situations, a highly desirable property.
+
+# Detailed design
+The basic infrastructure to add and remove individual replicas is already in
+place; the etcd raft implementation has built-in support for the integrity-safe
+addition or removal of single replicas, and that ability has already been
+exposed to our replicas via the ChangeReplicas function.
+
+The replicate queue already adds replicas to under-replicated ranges using
+ChangeReplicas (this is how the initial replicas are created on the second and
+third store of a three-node cluster).  The replicate queue does not currently
+remove ranges from over-replicated ranges; however, this functionality can be
+added with relatively few changes.
+
+## Code Changes
+The Replicate Queue will be augmented in the following ways:
+
+#### replicateQueue.needsReplication
+The existing `replicateQueue.needsReplication()` method will be modified to return
+`true` in the case where there are too many replicas. This method also returns
+the difference between the current number of replicas and the target number;
+this will be negative in the case of too many replicas.
+
+`replicateQueue.shouldQueue()` is currently using the difference returned by
+`needsReplication()` as the priority for the replica; the method will now return
+the absolute value of the difference.
+
+#### allocator
+The `allocator` class is currently used by the replicate queue to select stores
+as replica targets.
+
+The class currently shares some scattered but similar logic between functions
+`allocateTargetInternal()`, `RebalanceTarget()` and `ShouldRebalance()` which
+determine how a given store's used capacity relates to the mean used capacity
+for the cluster.
+
+This code will be refactored into three functions:
++ `bestStore()`, which will return the "best" store in a slice of stores according to the
+  capacity criteria.
++ `worstStore()`, which returns the "worst" store in a slice.
++ `evaluateStore()` function, which will simply determine if a store is "good"
+  or "bad" according to the overall state of the cluster.
+
+The functionality within allocator will be refactored to use these functions.
+
+#### replicateQueue.process
+Process will now use the following logic for a replica which requires a replication change:
+
+1. If the range needs additional replicas:
+  + The queue will use the allocator's `AllocateTarget()` method to select a target node.
+  + If a target node is found, the queue will attempt to add a replica to it.
+2. If the range has too many replicas:
+  + The queue will grade the existing replicas using `allocator.worstStore()`.
+  + The worst replica will be removed.
+3. In either case, the queued replica will be re-queued to check for additional
+   changes, unless it was the replica that was removed.
+
+## Distribution Concerns
+The replication queue process can occur on any replica of a range; there is no
+requirement that it runs only on the leader or any other specific replica.
```

Comment:
`replicateQueue` only runs on the node with the leader lease. However, if the lease changes hands we do not abort any existing ChangeReplicas operations so we still need to be concerned about concurrency.